nexus/reconfigurator/planning/src/mgs_updates/mod.rs

```diff
@@ -327,6 +386,120 @@ fn mgs_update_status_sp(
     }
 }
 
+#[allow(clippy::too_many_arguments)]
+fn mgs_update_status_rot(
+    desired_version: &ArtifactVersion,
+    expected_active_slot: &ExpectedActiveRotSlot,
+    expected_inactive_version: &ExpectedVersion,
+    expected_persistent_boot_preference: &RotSlot,
+    expected_pending_persistent_boot_preference: &Option<RotSlot>,
+    expected_transient_boot_preference: &Option<RotSlot>,
+    found_active_slot: &ExpectedActiveRotSlot,
+    found_inactive_version: Option<&str>,
+    found_persistent_boot_preference: &RotSlot,
+    found_pending_persistent_boot_preference: &Option<RotSlot>,
+    found_transient_boot_preference: &Option<RotSlot>,
+) -> MgsUpdateStatus {
+    if &found_active_slot.version() == desired_version {
+        // If we find the desired version in the active slot, we're done.
+        return MgsUpdateStatus::Done;
+    }
+
+    // The update hasn't completed.
+    //
+    // Check to make sure the contents of the active slot, persistent boot
+    // preference, pending persistent boot preference, and transient boot
+    // preference are still what they were when we configured this update.
+    // If not, then this update cannot proceed as currently configured.
+    // It will fail its precondition check.
+    if found_active_slot.version() != expected_active_slot.version() {
+        return MgsUpdateStatus::Impossible;
+    }
+
+    if found_persistent_boot_preference != expected_persistent_boot_preference {
+        return MgsUpdateStatus::Impossible;
+    }
+
+    if found_pending_persistent_boot_preference
+        != expected_pending_persistent_boot_preference
+    {
+        return MgsUpdateStatus::Impossible;
+    }
+
+    if found_transient_boot_preference != expected_transient_boot_preference {
+        return MgsUpdateStatus::Impossible;
+    }
+
+    // If either found pending persistent boot preference or found transient
+    // boot preference are not empty, then an update is not done
+    if found_pending_persistent_boot_preference.is_some()
+        || found_transient_boot_preference.is_some()
+    {
+        return MgsUpdateStatus::NotDone;
+    }
+
+    // If there is a mismatch between the found persistent boot preference
+    // and the found active slot then the update is not done.
+    //
+    // TODO: Alternatively, this could also mean a failed update. See
+    // https://github.com/oxidecomputer/omicron/issues/8414 for context
+    // about when we'll be able to know whether an it's an ongoing update
+    // or an RoT in a failed state.
+    if found_persistent_boot_preference != &found_active_slot.slot {
+        // TODO-K: I am not 100% sure on this one. It may be impossible?
+        return MgsUpdateStatus::NotDone;
+    }
```

Comment:
I feel like this case, like the one above, should produce an error that causes the update process to stop (and eventually alert an operator).

Reply:
Hmm.  I thought the process was:

- say it's currently running out of slot A
- upload the image to slot B
- set _pending_ persistent boot preference to that slot (B) -- persistent boot preference would still be A
- reset device
- assuming things worked, now persistent boot preference is B _and_ active slot is B

and I wouldn't have expected to see a point where the active slot didn't match the persistent boot preference.  Is that wrong?  (Side note: do we have a description somewhere of the values of all these fields across an update?)